README
for using justify you have to enable html-tags in Dokuwiki

fckg/Changes
rev 11-Sep_14-19_31

Note:  For the log of Changes prior to thei document, see the file named Changes.1.

11 May 2011
A completely new handling of the Dokuwiki draft mechanism.  The previous implementation
for accessing drafts was out-of-date.  These changes meant pruning away a great deal of 
unnecssary code (and there is always the chance that an error crept in when pruning).

12 May 2011
Improved Ajax handling for re-naming DW's draft files to fckgLite names, since not
all browsers handled the previous method successfully.  Current method runs an
interval timer to check for DW draft files, whereas previously the Ajax call was made
from the timer function in handler.php.

Changed regex for handling angle brackets inside code blocks (edit.php), using
entities instead of "< " and " >", &lt; and &gt;.

Added assignment of FCKeditor's Editor Node to helper.php's FCKeditor_OnComplete();
fckeditor.html also makes this assigment but has to keep re-trying until is succeeds
in connecting to FCK.EditorDocument, whereas this method is more immediate. This
assigment is critical to the draft system.

13 May 2011
Decreased draft time from minute to 45 seconds (DW writes draft every 30 secs).
Decreased time for renaming of DW draft files to fckgLite draft files.
Added lock to renaming of DW draft files in scripts/draft_rename.php.

19 May 2011
Removed "TEST" designation from the distribution which offers a new implementation of the
Dokuwiki Draft mechanism.

19  May 
Removed DW Renderer's acronym processing from the output to the FCKEditor editor;
this does not interfere with acronyms processed for final output in browser. This
solves a problem of acronymic names that appear in plugins and plugin syntax, since
the editor sess these as acronyms instead of as plugin syntax.
 
23 May
Fixed ACL check in filemanager/connectors/php/check_acl.php which was not fully
backwardly compatible with DW after upgrading %USER% wildcard substitions and which
could fail under some conditions even in current DW distros. 

Put array check into edit.php when checking for guest user in edit.php where
the save and delete buttons are disabled for guests (approx lines 435ff.).  Some
systems got a PHP warning because the $INFO[userinfo][grps] did not hold an array.

25 May
Forward slashes in userewrite=2 were not being converted to colons, when saved.
This is now fixed in edit.php.

3 June
Re-worked the new draft recovery facility so that fckl draft files are removed in a
timely fashion, so as to prevent false draft file notifications.  Save, Show, Delete,
and Cancel buttons call draft_rename(), which deletes the fckl file, so that
it remains on the server only in cases where there is some interruption of
normal processing.

4 June
Made Call to draft_rename synchronous; otherwise there is no guarantee that it will
delete the draft file from the server.

5 June
Renamed draft_refresh to draft_delete in function calls and fckg/scripts to reflect
the new use of that function (see 3 June).

6 June
Added regex to save.php that preserves %[a-f]{2} sequences, like %ad in %admin,
from being mistaken by rawurldecode for html-encoded characters.

Cleaned up some useless code in meta.php.

Added check in edit.php to ignore draft files left over by DW Edit previews
when switching from DW Edit to FCKG Edit; these are not deleted, as they
are when saving of canceling when in DW Edit.


7 June
Fixed problem where HTML brackets were replaced by entities in GESHI blocks. Open brackets
are now replaced by "< ", and the space is removed on save. 

9 June
Permanent fix for draft file left over after DW Edit preview -- there was a bug in the
Javascript call, which corrupted  draft filename.

10 June
Spoke too soon on 9 June.  With multiple previews in DW Edit, draft file remains on
server.  This is now removed in meta.php set_session().

Added additional check to above by creating a new POST/hidden input item for the edit form
Qbject that appears in the $_REQUEST object only when DW Edit preview is in effect.

20 June
Inserted paragraph break above a code block that comes at top of file to enable insertion
of text above it; it is removed on save.

25 June
Added mechanism to handle swf files.

26 June
Fix for corrupted code blocks containing MULTI_PLUGIN (uses old style "< " and " >' instead
of &lt; and &gt; for html-style elements.

Added flash icon to FCKeditor

Added %%. . .%% escape around characters, which is converted o <nowiki>; this is an extension
of the earlier use of this on 12 Jan 2011.

27 June
Added support for nolinks and direct image links.

28 June
Fixed bug where direct (internal) links were being identified as external; also where
external direct were being missed.

In link.js removed wikilink class from URL when re-writing a url that had been an internal link.

30 June
Upgraded ent.ser to handle addtional Latin entities.

4 Jul
Fix for bugs in converting percent escapes to nowiki tags

Further Fix:  created single preg_replace_callback which handles all nested configurations of
percents and nowiki tags.

15 Jul
In helper.php: fixed bug in passing usergroups to filebrowser's browser.html, which affects
acl determination in config.php

17 Jul
In helper.php: intercept DW's onbeforeunload to issue our own warning when changes have
been made but not saved.

20 Jul 
Added support for link-only images.  Experimental.

Set window.dwfckTextChanged to false on save, because Anteater gives onbeforunload warning
on save.

21 Jul
Removed onbeforeunload from Safari which freezes when message is canceled and improved
keyhandler install.

1 Aug
Removed upload and folder create buttons from filebrowser when DW's 'Show pagesource' is in effect.

Eliminated php warning from an implode in helper.php.

12 Aug
aspell.php: added isset() check for 'FCK_aspell' index in $COOKIE array.

2 Sep
fckeditor link dialog:  when manually entering internal links, any link name that is not specified 

fckg/action/meta.php
<?php
/**
 *
 */
 
if(!defined('DOKU_INC')) die();
if(!defined('DOKU_PLUGIN')) define('DOKU_PLUGIN',DOKU_INC.'lib/plugins/');
require_once(DOKU_PLUGIN.'action.php');
global $conf;
global $fckg_lang;


$default_english_file = DOKU_PLUGIN . 'fckg/action/lang/en.php';
require_once($default_english_file);
if(isset($conf['lang'])) {
  $default_lang_file = DOKU_PLUGIN . 'fckg/action/lang/' . $conf['lang'] . '.php';
  if(file_exists($default_lang_file)) {                                       
    @include_once($default_lang_file);
  }
}
 
class action_plugin_fckg_meta extends DokuWiki_Action_Plugin {
  var $session_id = false;    
  var $draft_file;
  
  /**
   * return some info
   */
  function getInfo(){
    return array(
		 'author' => 'Myron Turner',
		 'email'  => 'turnermm02@shaw.ca',
		 'date'   => '2007-09-18',
		 'name'   => 'meta',
		 'desc'   => 'fckg  meta handler',
		 'url'    => 'http://www.mturner.org/fckgLite',
		 );
  }
 

  /*
   * Register its handlers with the dokuwiki's event controller
   */
  function register(&$controller) {            
            $controller->register_hook( 'TPL_METAHEADER_OUTPUT', 'AFTER', $this, 'loadScript');    
            $controller->register_hook( 'HTML_EDITFORM_INJECTION', 'AFTER', $this, 'preprocess'); 
            $controller->register_hook( 'HTML_EDITFORM_OUTPUT', 'BEFORE', $this, 'insertFormElement');            
            $controller->register_hook('DOKUWIKI_STARTED', 'BEFORE', $this, 'file_type');         
            $controller->register_hook('TPL_CONTENT_DISPLAY', 'AFTER', $this, 'prevent_output');       
            $controller->register_hook('DOKUWIKI_STARTED', 'AFTER', $this, 'fnencode_check');                 
  }

 
 function  insertFormElement(&$event, $param) {	 
   global $FCKG_show_preview;  

  $param = array();

   global $ID;
   $dwedit_ns = @$this->getConf('dwedit_ns');
   if(isset($dwedit_ns) && $dwedit_ns) {
       $ns_choices = explode(',',$dwedit_ns);
       foreach($ns_choices as $ns) {
         $ns = trim($ns);
         if(preg_match("/$ns/",$ID)) {
            echo "<style type = 'text/css'>#edbtn__preview,#edbtn__save, #edbtn__save { display: inline; } </style>";         
            break;
         }
       }
   }
   $act = $event->data;
   if(is_string($act) && $act != 'edit') {  
        return;
   }

  // restore preview button if standard DW editor is in place
  // $FCKG_show_preview is set in edit.php in the register() function
 if($_REQUEST['fck_preview_mode'] != 'nil' && !isset($_COOKIE['FCKW_USE'])) {    
     echo '<style type="text/css">#edbtn__preview { display:none; }</style>';
 }
 elseif($FCKG_show_preview) {
      echo '<style type="text/css">#edbtn__preview { display: inline; } </style>';
 }
 else {
    echo '<style type="text/css">#edbtn__preview, .btn_show { position:absolute; visibility:hidden; }</style>';
 }
  
 global $fckg_lang;

  if($_REQUEST['fck_preview_mode']== 'preview'){
    return;
  }

 $param = array();
 $this->preprocess($event, $param);  // create the setDWEditCookie() js function
 $button = array
        (
            '_elem' => 'button',
            'type' => 'submit',
            '_action' => 'cancel',
            'value' => $fckg_lang['btn_fck_edit'],
            'class' => 'button',
            'id' => 'edbtn__edit',            
            'title' => $fckg_lang['btn_fck_edit']             
        );

     $pos = strpos($_SERVER['HTTP_USER_AGENT'], 'MSIE');
     if($pos === false) {
                 $button['onclick'] = 'return setDWEditCookie(1, this);';
     }
     else {
                $button['onmousedown'] = 'return setDWEditCookie(1, this);';
     }

    $pos = $event->data->findElementByAttribute('type','submit');
       //inserts HTML data after that position.
    $event->data->insertElement(++$pos,$button);

   return;
 
  }


 function preprocess(&$event, $param) {	 
    $act = $event->data;
 
   if(is_string($act) && $act != 'edit') {  
        return;
   }
  global $INFO;
  $cname =  $INFO['draft'];   
    
 $url = DOKU_URL . 'lib/plugins/fckg/scripts/jq_alt.js';    
 
  echo <<<SCRIPT
    <script type="text/javascript">
    //<![CDATA[ 
    
      if(!window.jQuery){
        LoadScript("$url"); 
      }
 
      
   
    
    function setDWEditCookie(which, e) { 
       var cname = "$cname";       
       var dom = document.getElementById('fckg_mode_type');          
       if(which == 1) {
           dwedit_draft_delete("$cname");
           document.cookie='FCKW_USE=other;expires=Thu,01-Jan-70 00:00:01 GMT;'
           if(e && e.form) {
                    if(e.form['mode']) {
                       e.form['mode'].value = 'fck';
                    }
                    else {
                       e.form['mode'] = new Object;
                       e.form['mode'].value = 'fck';  
                    }
           }
           else dom.value = 'fck';  
       }
        else {            
            var nextFCKyear=new Date();
            nextFCKyear.setFullYear(nextFCKyear.getFullYear() +1 );
            document.cookie = 'FCKW_USE=_false_;expires=' + nextFCKyear.toGMTString() + ';';    
            dom.value = 'dwiki';        

        }
         e.form.submit();
    }
  
 
    //]]> 

    </script>
SCRIPT;

  }


  function set_session() {	
      global $USERINFO, $INFO; 
      global $conf; 
      global $ID;
      global $ACT;

      if($this->session_id) return;       

           $cname = getCacheName($INFO['client'].$ID,'.draft');  
           $fckl_draft = $cname . '.fckl';
           if((isset($ACT) && is_array($ACT)) || isset($_REQUEST['dwedit_preview'])) {
              if(isset($ACT['draftdel']) || isset($ACT['cancel']) || isset($_REQUEST['dwedit_preview'])) {
                 @unlink($fckl_draft);   
                 @unlink($cname); 
              }
           }

           if(file_exists($cname)) {
              if(file_exists($fckl_draft)) {
                    unlink($fckl_draft);
              }
              @rename($cname, $fckl_draft);
           }

          
           $session_string =  session_id(); 
           $this->session_id = $session_string;      
       

           $_SESSION['dwfck_id'] = $session_string; 
           $default_fb = $this->getConf('default_fb');

           if($default_fb == 'none') {
               $acl = 255; 
           }
           else {
              $acl = auth_quickaclcheck($ID); 
           }
           $_SESSION['dwfck_acl'] = $acl; 

           if($this->getConf('openfb') || $acl == 255) {
             $_SESSION['dwfck_openfb'] = 'y';
           }
           else { 
              $_SESSION['dwfck_openfb'] = 'n';
           }

           $_SESSION['dwfck_grps'] = $USERINFO['grps'];
           $_SESSION['dwfck_client'] = $INFO['client'];   
           $_SESSION['dwfck_sepchar'] = $conf['sepchar'] ;   
           $_SESSION['dwfck_conf'] = array('sepchar'=> $conf['sepchar'],
                  'isadmin'=>($INFO['isadmin'] || $INFO['ismanager']), 
                  'deaccent'=>$conf['deaccent'], 'useslash'=>$conf['useslash']);
           $elems = explode(':', $ID);  
           array_pop($elems);
 
           $_SESSION['dwfck_ns'] = implode(':',$elems);        
           $_SESSION['dwfck_top'] = implode('/',$elems);           
           $_SESSION['dwfck_del'] = $this->getConf('acl_del');
           
            // temp fix for puzzling encoding=url bug in frmresourceslist.html,
           // where image loading is processed in GetFileRowHtml()

           if(preg_match('/fckg:fckeditor:userfiles:image/',$ID)) {
                      $_SESSION['dwfck_ns'] = "";        
                      $_SESSION['dwfck_top'] = "";      

            }

          // $expire = time()+60*60*24*30;
           $expire = null;
           setcookie('FCK_NmSp_acl',$session_string, $expire, '/'); 
  
        
  }

  function file_type(&$event, $param) {	 
       global $ACT, $TEXT;
       global $USERINFO, $INFO, $ID; 

       if(isset($_COOKIE['FCK_NmSp'])) $this->set_session(); 
       /* set cookie to pass namespace to FCKeditor's media dialog */
      // $expire = time()+60*60*24*30;
       $expire = null;
       setcookie ('FCK_NmSp',$ID, $expire, '/');     
      
          

      /* Remove TopLevel cookie */         
       if(isset($_COOKIE['TopLevel'])) {
            setcookie("TopLevel", $_REQUEST['TopLevel'], time()-3600, '/');
       }

    
       if(!isset($_REQUEST['id']) || isset($ACT['preview'])) return;
       if(isset($_REQUEST['do']) && isset($_REQUEST['do']['edit'])) {
              $_REQUEST['do'] = 'edit';
       }
  } 

function loadScript(&$event) {
  echo <<<SCRIPT

    <script type="text/javascript">
    //<![CDATA[ 
    function LoadScript( url )
    {
     document.write( '<scr' + 'ipt type="text/javascript" src="' + url + '"><\/scr' + 'ipt>' ) ;        

    }
//]]> 

 </script>

SCRIPT;

}
/*
function fck_editor(&$event) {
  
    if($_REQUEST['fckEditor_text'] ) {
        echo "<div id = 'fckEd__text' style='position:absolute; visibility:hidden;'>";   
        echo $_REQUEST['fckEditor_text'];     
        echo '</div>';
    }


}
*/

/** 
 *  Handle features need for DW Edit: 
 *    1. load script, if not loaded
 *    2. Re-label Cancel Button "Exit" when doing a preview  
*/
  function prevent_output(&$event) {
  global $ACT;

  $url = DOKU_URL . 'lib/plugins/fckg/scripts/script-cmpr.js';
  echo <<<SCRIPT

    <script type="text/javascript">
    //<![CDATA[ 

    try {
    if(!HTMLParserInstalled){
         LoadScript("$url");        
    }
    }
    catch (ex) {  
         LoadScript("$url");        
    }             
    function createRequestValue() {
        try{
        var inputNode=document.createElement('input');
        inputNode.setAttribute('type','hidden');
        inputNode.setAttribute('value','yes');
        inputNode.setAttribute('name','dwedit_preview');
        inputNode.setAttribute('id','dwedit_preview');
        var dwform = GetE("dw__editform");
        dwform.appendChild(inputNode);
        }catch(e) { alert(e); }
    }
//]]> 
 </script>

SCRIPT;

  if(isset($_REQUEST['do']) && is_array($_REQUEST['do'])) {
    if(isset($_REQUEST['do']['preview'])) {
           echo '<script type="text/javascript">';
           echo ' var dwform = GetE("dw__editform"); dwform["do[draftdel]"].value = "Exit"';
           echo "\ncreateRequestValue()\n";
           echo  '</script>';
    }
  }


  }

function fnencode_check() {


       global $conf;
       global $updateVersion;
       $rencode = false;
	  
        if($conf['fnencode'] != 'safe') return;
		if (file_exists(DOKU_PLUGIN . 'fckg/conf/saferencode.revert')) {
		     return;
		}
        if(isset($updateVersion) && $updateVersion >= 31) {           
          $rencode = true;     
        }
        else {
            $list = plugin_list('action');
            if(in_array('safefnrecode', $list)){
                $rencode = true;   
     
            }
            elseif(file_exists($conf['datadir'].'_safefn.recoded') ||
               file_exists($conf['metadir'].'_safefn.recoded') ||
               file_exists($conf['mediadir'].'_safefn.recoded') )
            { 
               $rencode = true;
            }
        }


      if($rencode && !file_exists(DOKU_PLUGIN . 'fckg/saferencode')) {
         msg("This version of fckgLiteSafe does not support the re-encoded safe filenames. "
         . "You risk corrupting your file system.  Download an fnrencode version from either gitHub or the fckgLite web site."
         . " <a style='color:blue' href='http://www.dokuwiki.org/plugin:fckglite?&#fckglitesafe'>See fckgLite at Dokuwiki.org</a>  ",
            -1);
      }	 
      else if(!$rencode && file_exists(DOKU_PLUGIN . 'fckg/saferencode'))   {
	    msg("This version of fckgLiteSafe requires a newer version of Dokuwiki (2011-05-25 Rincewind or later).  You risk corrupting your file system. "
		 .   "To convert this distribution of fckgLite/fckgLiteSafe for use with earlier versions of Dokuwiki,  see the README file or " 
		 . " <a style='color:blue' href='http://www.mturner.org/dwfck/fckgLite/doku.php?id=docs:upgrade_6&#anteater'>or the fckgLite web site</a>  ",
		-1);
    }	  
}




      


function write_debug($data) {
  return;
  if (!$handle = fopen(DOKU_INC .'meta.txt', 'a')) {
    return;
    }
  if(is_array($data)) {
     $data = print_r($data,true);
  }
    // Write $somecontent to our opened file.
    fwrite($handle, "$data\n");
    fclose($handle);

}

}



